src/meepgeom.cpp

```diff
@@ -270,13 +280,116 @@ bool is_metal(meep::field_type ft, const material_type *material) {
   else
     switch (md->which_subclass) {
       case material_data::MEDIUM:
+      case material_data::MATERIAL_GRID:
         return (md->medium.mu_diag.x < 0 || md->medium.mu_diag.y < 0 || md->medium.mu_diag.z < 0);
       case material_data::PERFECT_METAL:
         return false; // is an electric conductor, but not a magnetic conductor
       default: meep::abort("unknown material type"); return false;
     }
 }
 
+meep::realnum material_grid_val(vector3 p, material_data *md) {
+  // given the relative location, p, interpolate the material grid point.
+
+  if (!is_material_grid(md)) { meep::abort("Invalid material grid detected.\n");}
+  meep::realnum u;
+  u = meep::linear_interpolate(p.x, p.y, p.z, md->design_parameters, md->grid_size.x,md->grid_size.y, md->grid_size.z, 1);
+  return u;
+}
+
+static int matgrid_val_count = 0; /* cache for gradient calculation */
+meep::realnum matgrid_val(vector3 p, geom_box_tree tp, int oi, material_data *md)
+{
+  meep::realnum uprod = 1.0, umin = 1.0, usum = 0.0, udefault = 0.0, u;
+  matgrid_val_count = 0;
+  
+  // iterate through object tree at current point
+  if (tp) {
+    do {
+        u = material_grid_val(to_geom_box_coords(p, &tp->objects[oi]),(material_data *)tp->objects[oi].o->material);
+        if (matgrid_val_count == 0) udefault = u;
+        if (u < umin) umin = u;
```

Comment:
I'm following the MPB convention by using `to_geom_box_coords`, rather than the more general `to_geom_object_coords`. When we enable polygons for material grids, we should consider this.

Reply:
`to_geom_box_coords` just calls `to_geom_object_coords` after shifting the center…

Reply:
Since we have an `geom_box_object` taken from a tree, I think `to_geom_box_coords` is the right thing.